Remove commit from internal wallet

ffakenz opened the pull request:
<!-- Describe your change here -->
Closes #954 

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Introduce new strategy for committing in BehaviourSpec

We added a field `simulateCommit` in the `SimulatedChainNetwork` in order
to easily _post_ a commit tx to the chain directly. The test that assert we
see a `CommandFailed` are deleted since we can't expect to see this once the
client command for committing is removed.

v0d1ch committed:
Tackle the model spec to do a commit

We used _injectChainEvent_ from `TestHydraClient` to inject
the commit tx to all of the existing nodes.

ffakenz committed:
Fix golden specs

ffakenz committed:
Remove Commit from the api and logs yaml files

ffakenz committed:
Update CHANGELOG

v0d1ch committed:
Remove now unneeded Marked type

v0d1ch committed:
Remove PostTxError constructors related to wallet

We have already ErrCoverFee type to hold all wallet errors.
Instead of mapping those to different constructors in PostTxError
why don't we have only InternalWalletError that would wrap the
ErrCoverFee and also provide fields for utxo and transaction.
Left a TODO to add To/From json for ErrCoverFee so that we can
have the nicer output.

v0d1ch committed:
Remove datum marking

- Delete related functions and types (markerDatum, isMarkedDatum, OutputMarkerHash).
- Alter the docs to reflect on this.
- Remove option for marker-hash in hydra-tools executable.
- Drop the wallet test that are no longer related.
- Rename findFuelOrLargestUTxO to findLargestUTxO.

v0d1ch committed:
Remove fuel-testnets.sh and fix documentation

v0d1ch committed:
Rename ErrNoFuelUTxOFound to ErrNoUTxOFound

v0d1ch committed:
Propagate fuel removal to hydra-tui

v0d1ch committed:
Remove queryMarkedUTxO

Propagate the changes further removing the tracing of the fuel utxo
completely.

ch1bo commented on testnets/fuel-testnet.sh:
There are other scripts still referring to the `fuelMarker` which need to be cleaned up as well.

ch1bo commented on testnets/fuel-testnet.sh:
Minor: the demo/seed-devnet.sh script still mentions fuel marking as well (although does not do it)

ch1bo commented on hydra-cluster/test/Test/DirectChainSpec.hs:
```diff
@@ -408,7 +407,7 @@ spec = around showLogsOnFailure $ do
             -- right reason here (see corresponding mutation test for this).
             contestAgain `shouldThrow` \case
-              (ScriptFailedInWallet{} :: PostTxError Tx) -> True
+              (InternalWalletError{} :: PostTxError Tx) -> True
```
Why did this exception change? The previous exception was much more concrete (judging from the note above)

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -223,23 +223,6 @@ components:
             type: string
             enum: ["Abort"]
 
-    Commit:
```
Must: remove the entry for this in the `clientInput` operation [above](https://github.com/input-output-hk/hydra/blob/f0655942bde4148ff59b79bb399e9c00b17cda8c/hydra-node/json-schemas/api.yaml#L105)

Should: update the state diagram in the [docs](https://github.com/input-output-hk/hydra/tree/f0655942bde4148ff59b79bb399e9c00b17cda8c/docs/core-concepts/behavior.md) to not have a `Commit` event/client input, but instead mention that a commit transaction needs to be built + submitted (maybe even refer `POST /commit`)

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -134,16 +134,8 @@ data PostTxError tx
-    -- here. But the wallet still uses ledger types and we don't want to copy the
-    -- conversion from ledger 'TransactionScriptFailure' to the cardano-api
-    -- 'ScriptExecutionError' type.
-    ScriptFailedInWallet {redeemerPtr :: Text, failureReason :: Text}
```
Why did you remove these errors?

Especially the errors for `NotEnoughFuel` or `NoFuelUTxOFound` are meaningful to the clients and require user action!

Also, the removal of `ScriptFailedInWallet` seems very unrelated to the goal of this PR

v0d1ch replied:
I've removed them since I wanted to get rid of `PostTxError` being a kitchen sink for different kinds of errors. Let's design a better solution then, perhaps in another PR.

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -134,16 +134,8 @@ data PostTxError tx
-    ScriptFailedInWallet {redeemerPtr :: Text, failureReason :: Text}
   | -- | A generic error happened when finalizing a transction in the wallet.
-    InternalWalletError {headUTxO :: UTxOType tx, reason :: Text, tx :: tx}
+    InternalWalletError {relatedUtxo :: UTxOType tx, reason :: Text, tx :: tx}
```
Must: update `api.yaml` if these fields are changed.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -215,12 +213,15 @@ getTxId tx = Ledger.TxId $ SafeHash.hashAnnotated (body tx)
 
+-- TODO: provide To/From JSON instances for ErrCoverFee so that we can reuse
+-- them in InternalWalletError instead of putting just the Text for reason.
+--
```
I would strongly advise against this. The `ErrCoverFee` type is not intended to be used in an interface and it's typically a good idea to constrain de-/serialization to the "edges" of a system. We have `PostTxError` already which can/should be used for this. Or even better, one or more dedicated error types in `Hydra.Api...` with serialization instances should be used.

v0d1ch replied:
Ok, let's do it differently then https://github.com/input-output-hk/hydra/pull/1018/files#r1294541076

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
Should: also remove the `PostTx(CommitTx)` data constructor (line 69)

This is now unused and might allow to remove more code from the `hydra-node`, resulting in only one way to construct commit transactions now (via the `draftCommitTx` function and not anymore through `postTx`)

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -645,6 +615,17 @@ simulatedChainAndNetwork initialChainState = do
+          { observedTx = OnCommitTx{party, committed}
+          , newChainState = cs'
+          }
+    recordAndYieldEvent nodes history chainEvent
```
Should: DRY with `postTx` above.

The act of simulating observation of a Hydra protocol transaction (either `OnCommitTx` or the result of `toOnChainTx`) is identical and can be re-used between `postTx` and `simulateCommit`

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -555,7 +555,12 @@ performCommit parties party paymentUTxO = do
+          Observation
+            { observedTx = OnCommitTx{party, committed = realUTxO}
+            , newChainState = initialChainState
+            }
```
Should: use `simulateCommit` instead

This is requiring maybe some refactoring, but the Model spec code should follow the Behavior spec. Also, it would be an ill-formed chain event with the new chain state being not consistent (`initialChainState` creates an `Idle` chain state)

v0d1ch replied:
I've tried but it is not so easy. My initial concern was also that the state we are passing is just the starting one and feels like the model spec is wrongfully asserting everything is fine since we are not tracking the state. Curious to see how to get a hold of `SimulatedChainAndNetwork` inside of `performCommit` for example.

ch1bo commented on hydra-tui/src/Hydra/TUI.hs:
```diff
@@ -369,10 +369,6 @@ handleAppEvent s = \case
-      NotEnoughFuel ->
-        s
-          & warn' time "Not enough Fuel. Please provide more to the internal wallet and try again."
-          & stopPending
```
This would have been a good hint to _not_ remove `NotEnoughFuel` as it provides clear instructions to the user (much better than the `InternalWalletError` case below)

ch1bo reviewed (changes requested):
Overall looks good, but requires some more changes.

Also, please grep for `"marker"` in the code base to find some more mentions to purge from the docs (glossary) and some comments in scripts.

v0d1ch committed:
Remove marker mention from squash-utxo script and glossary

v0d1ch committed:
Remove fuel wording in seed-devnet script

v0d1ch committed:
Fix the uml diagram and explain now missing Commit client input

v0d1ch committed:
Remove outdated Commit message client input

v0d1ch committed:
Bring back the old error types

v0d1ch committed:
Dry code in BehaviorSpec

v0d1ch committed:
WIP - Implement successfully draftCommitTx in DirectChainSpec

- Leftover work is to remove CommmitTx and replace it with the
external commit workflow demonstrated in this commit.

v0d1ch committed:
Fix golden files again to start from a green ci

v0d1ch committed:
Add SimulatedChainNetwork to the RunMonad state

ffakenz committed:
draft simulateCommit

v0d1ch committed:
Create a helper function to mimic external commit

- We will need to do external commit often in DirectChainSpec so
helper function would help us out.

- Also remove error call if draftCommitTx fails in test code and use
  throwIO to be able to assert certain exceptions are thrown in tests.

v0d1ch committed:
Remove CommitTx

Fix all occurrences of CommitTx in codebase and
make the DirectChainSpec tests green

v0d1ch committed:
Fix golden files and some wording inside of quickstart

ffakenz committed:
Remove Commit from api spec

This fixes docs validate-api script.

ch1bo commented on docs/core-concepts/behavior.md:
```diff
@@ -6,9 +6,9 @@ sidebar_position: 4
 This page documents the behavior of a `hydra-node` at the API layer. That is, how the system behaves given [ClientInputs](pathname:///haddock/hydra-node/Hydra-API-ClientInput.html#t:ClientInput) and what [ServerOutputs](pathname:///haddock/hydra-node/Hydra-API-ServerOutput.html#t:ServerOutput) are produced in response to it. See also the [API reference](/api-reference/) for more details about ind ...
 
-The formalism uses [UML statechart](https://en.wikipedia.org/wiki/UML_state_machine) language where transitions are labeled: `input [condition] / output`. When two outputs (e.g. `A` and `B`) are expected we write `A,B`, while `{A,B}` denotes mutual exclusiveness of outputs.
+The formalism uses [UML statechart](https://en.wikipedia.org/wiki/UML_state_machine) language where transitions are labeled: `input [condition] / output`. When two outputs (e.g. `A` and `B`) are expected we write `A,B`, while `{A,B}` denotes mutual exclusiveness of outputs. The only discrepancy is http `POST /commit` action which is not a state transition but a user action that submits a commit t ...
```
I would suggest to move the last sentence one paragraph up which speaks about the Client inputs and outputs.

In terms of UML state diagrams, inputs can be anything.. so it's perfectly fine that an HTTP request is mixed with the websocket message received.

ch1bo commented on docs/core-concepts/behavior.md:
```diff
@@ -6,9 +6,9 @@ sidebar_position: 4
-![](https://www.plantuml.com/plantuml/svg/ZP4_J_Cm48Vt-nIUNpLwzmweA18Y2o28MAeEBdUX9zddABv02kAxOzVGG65Wkp_FSr5-NaMrWuxUmOwH3FbUzmOfFWpSAmS1MF_RcAewCusmidNyml9ebeVM_3UtP77VXZfupmhm3Vef5InffL324oCf5opM9VPy6uQCNfB59kRkL_ow9qdq9vTRWCDmNbxHSpibvBMTxfEBYtPgv2bNRuixiNtS1Qs3T3numYBdqxKBaT0iIt7yH1a3VAo_W3CIH2aguH7AFsObZ7eJOkIjeqIlbJsPtiP8qTXEQJ2OTCT2cpdQeW6Sw9MZJkUuolKvhpJEAH425ABmVtl65GcEthq3)
+![](https://www.plantuml.com/plantuml/svg/ZP71JW8n48RlVOevge5mvs0mH2CN8RBnGZWKEiWaxNJJ3hfWV7VRBIw87hnrPhvl_-vq54K7sJchjcGGqDMo1uDn7QWMygpKucO9_VujJ9Y4jAK3yIiCsn86y8pQx2i_ziwHAFK3-YrTpQRp2WRhbhvEUl44pOMPr0TYRPDpj_8X9pscf4dCrP_uj4PEz3UNIwNQvcduXEzLav2Fgdb9hkbLpOJVZgVxfgQ0vhCtPrt7hPUnvmq5XwPy9eUChOzeO5WENLXfAtKSduCTubam2feEoh-esUzavcEabSL4BuGSGgrZn0Xw8nZ09DqIu_AqxA8fTQ7tBMxaR75btDsWRTCXtxCGXV_VmuwGp ...
 
 [Edit this diagram](https://www.plantuml.com/plantuml/uml/ZP6zJWCn383tF8Ldr8gz0oggG7G112UMgaEANMefSKw92zSYtftaKbrBOM3pp_UTVSuqgOswzSvi60d8jxe3fFGQkLKEOipYOWdZyHNCXPMjuptB6qpw52xXYIpkcZ0BthCQymFwBLKiQpLGmZCZAxSircNsUXLYGU_8uZoLx4_yeIM1oS2Lr9Y-U6pUqqVJPUIpwLwoYwEccylx8RhfMew4NwDdiMtQ19q5MNFqXFKpPSZCXKG8aneCmshPc4Fx51oG84f92GUe_AALiDN7a1Al76LweUm9MfiI9R1hfGYOTST2o-EHgsRCcJvsjDQJmzNsi45VvdtPinAL_z-3J ...
```
Should: also update the edit link to include the latest changes, otherwise editing would start with the old diagram.

ch1bo commented on hydra-cluster/test/Test/DirectChainSpec.hs:
```diff
@@ -479,3 +495,21 @@ delayUntil :: (MonadDelay m, MonadTime m) => UTCTime -> m ()
+
+  commitTx <- draftCommitTx utxoToCommit
+  let signedTx = signTx externalSk commitTx
+  void $ submitTransaction networkId nodeSocket signedTx
```
Could: use `submitTx` which we created for the EndToEndSpec and takes a `RunningNode`

ch1bo commented on hydra-cluster/test/Test/DirectChainSpec.hs:
```diff
@@ -433,16 +445,20 @@ withDirectChainTest ::
+            eTx <- draftCommitTx utxo
+            case eTx of
+              Left e -> throwIO e
+              Right tx -> pure tx
```
Observation: No need to change here, but interesting that you map the "expected errors" to "exceptional errors" here. This is an indication of why we should have only `m Tx` here and not `m (Either e Tx)` (or only `Either e Tx` if possible).. it provides a more uniform interface.

v0d1ch replied:
I had to do it because one of the tests use `shouldThrow` to assert proper exception is thrown now that we actually  use `draftCommitUTxO` in these tests.

v0d1ch committed:
Rename external commit functions to reflect their real behavior

ch1bo commented on hydra-node/test/Hydra/Model/MockChain.hs:
```diff
@@ -91,6 +91,18 @@ mockChainAndNetwork tr seedKeys cp = do
+            hydraNodes
+            ( \MockHydraNode{node = HydraNode{eq = EventQueue{putEvent}}} ->
+                putEvent (OnChainEvent chainEvent)
+            )
```
Must: not implement it this way

The MockChain uses real blocks and the full chain layer (including it's local chain state tracking). That means, that we should not just inject events with `initialChainState` here. Instead, this should create a real commit transaction and put it into the mock chain, such that the chain handler can observe the transaction and infer the right chain state update from it.

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -163,8 +163,8 @@ waitForAll tracer delay nodes expected = do
-externalCommit' :: HydraClient -> UTxO' TxOutWithWitness -> IO Tx
-externalCommit' HydraClient{hydraNodeId} utxos =
+externalCommitViaApi' :: HydraClient -> UTxO' TxOutWithWitness -> IO Tx
+externalCommitViaApi' HydraClient{hydraNodeId} utxos =
```
Why did you rename this function?

v0d1ch replied:
I introduced `mimicExternalCommit` in the `DirectChainSpec` which is not optimal name for this function so I wanted to distinguish between these two where one uses hydra api and the other one just submits the tx directly.

ch1bo replied:
I agree that `externalCommit` is quite vague, and does not clarify what it's doing fully.

* Adding `ViaApi` here does not explain better what it does. The means how it does an external commit or draft the tx are frankly irrelevant to the caller of this function. It gives a `HydraClient` and expects a result.
* So maybe expanding on the verb is what we rather should do here. Looking at the haddocks better names (IMO) would be `draftCommitTx` or `createCommitTx` or `requestCommitTx`

v0d1ch replied:
`requestCommitTx` it is!

ch1bo reviewed (changes requested):
The BehaviorSpec / MockChain is taking a short cut still, but otherwise this looks good.

v0d1ch committed:
Use proper record when constructing Nodes

v0d1ch committed:
Submit a commitTx instead of using event

To be consistent on how we decided to post commit tx in the tests
(Behavior and Model) we don't want to just post an event - instead it
would be better to make the commit tx come from the chain directly.

v0d1ch committed:
Move sentence up in behavior.md

v0d1ch committed:
Fix uml edit link

v0d1ch committed:
Move the chain state creation back to connectNode

This caused memory leak and it was an artifact of working on
implementing simulateCommit but forgetting to revert the code.

ch1bo commented on hydra-tui/src/Hydra/Client.hs:
```diff
@@ -95,7 +95,7 @@ withClient Options{hydraNodeHost = Host{hostname, port}, cardanoSigningKey, card
     callback ClientDisconnected >> threadDelay 1 >> reconnect f
 
-  externalCommit' sk payload =
+  requestCommitTx' sk payload =
```
Could: This is not only requesting the commit tx, but actually does signing and submission as well, so we could name it more different than "just requesting"

v0d1ch replied:
This is a _victim_ of mass rename I did... reverting.

ch1bo reviewed (approved):
This looks good now! Just some :bike:-:house:-ing on names :)

v0d1ch committed:
Use submitTx instead of submitTransaction

v0d1ch committed:
Revert renaming of externalCommit -> requestCommitTx
